File: domain/use_case/execute_tool_use_case.py
from domain.tools.tool import Tool
from utils.logger import Logger


class ExecuteToolUseCase:
    def __init__(
        self,
        tools: dict[str, Tool],
        logger: Logger,
        working_directory: str,
    ):
        self._tools = tools
        self._logger = logger
        self._working_directory = working_directory

    def execute(self, tool_name: str, arguments: dict) -> str:
        self._logger.info(
            "Executing tool.",
            context={"tool_name": tool_name, "arguments": arguments},
        )
        try:
            tool = self._tools.get(tool_name)
            if not tool:
                raise ValueError(f"Tool '{tool_name}' not found.")

            arguments_copy = arguments.copy()
            arguments_copy["working_directory"] = self._working_directory

            result = tool.execute(**arguments_copy)

            self._logger.info(
                "Tool executed successfully.",
                context={"tool_name": tool_name, "result": result},
            )
            return result
        except Exception as e:
            self._logger.error(
                "Tool execution failed.",
                context={"tool_name": tool_name, "error": str(e)},
            )
            raise e

File: src/domain/use-case/execute-tool-use-case.ts
/**
 * Use case for executing tools.
 */

import type { Tool } from '../boundary/tool.js';
import type { Logger } from '../../utils/logger.js';

export class ExecuteToolUseCase {
  constructor(
    private readonly tools: Record<string, Tool>,
    private readonly logger: Logger,
    private readonly workingDirectory: string
  ) {}

  async execute(toolName: string, args: Record<string, unknown>): Promise<string> {
    this.logger.info('Executing tool.', {
      toolName,
      arguments: args,
    });

    try {
      const tool = this.tools[toolName];
      if (!tool) {
        throw new Error(`Tool '${toolName}' not found.`);
      }

      const argumentsCopy = { ...args };
      argumentsCopy.working_directory = this.workingDirectory;

      const result = await tool.execute(argumentsCopy);

      this.logger.info('Tool executed successfully.', {
        toolName,
        result,
      });

      return result;
    } catch (error) {
      this.logger.error('Tool execution failed.', {
        toolName,
        error: (error as Error).message,
      });
      throw error;
    }
  }
}
